ac.soton.eventb.emf.containment/model/containment.ecore

```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ecore:EPackage xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:ecore="http://www.eclipse.org/emf/2002/Ecore" name="containment" nsURI="http://soton.ac.uk/models/eventb/containment/2017"
+    nsPrefix="containment">
+  <eClassifiers xsi:type="ecore:EClass" name="ContainsExtension" eSuperTypes="../../org.eventb.emf.core/model/eventbcore.ecore#//AbstractExtension">
+    <eAnnotations source="org.eventb.emf.core.extendedMetaClasses" references="../../org.eventb.emf.core/model/eventbcore.ecore#//machine/Machine"/>
+    <eStructuralFeatures xsi:type="ecore:EReference" name="extension" eType="ecore:EClass ../../org.eventb.emf.core/model/eventbcore.ecore#//AbstractExtension"/>
+    <eStructuralFeatures xsi:type="ecore:EAttribute" name="names" upperBound="-1"
+        eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString" defaultValueLiteral=""/>
+  </eClassifiers>
+</ecore:EPackage>
```

Comment:
(For consideration only - could leave as is) 
The meta-model names feel a bit cumbersome & technical..
ContainsExtension -> Containment, or ContainedModel?
extension -> model?
names -> instances or instanceNames?

Reply:
I actually started with ContainedModel and changed to ContainsExtension. It is the extension that we want to add to the machine so I think extension is appropriate (and we include AbstractExtension anyway). The reason why it is not ContainedExtension is too be consistent with RefinesMachine and SeesContext. But maybe ContainedExtension is better.
The reason for using names is because I want it to be "element name" so that there cannot be name clash. However, It probably better to have a single name, instead of "names". In fact, I think I should try to write the XText grammar so that we can do this.

Reply:
but the names attribute is multiplicity many - doesn't this mean we can include multiple (named) instances of the extension?

Reply:
This is because in the XText grammar, we have
  contains
    a,b : ExtensionA
    c,d: ExtensionB
 which at the moment will create
   ContainsExtension{extension = ExtensionA, names=[a,b]}
   ContainsExtension{extension = ExtensionB, names=[c,d]}
 What is desirable (I'm working on it)
 is to have 
   ContainsExtension{name = a, extension = ExtensionA}
   ContainsExtension{name = b, extension = ExtensionA}
   ContainsExtension{name = c, extension = ExtensionB}
   ContainsExtension{name = d, extension = ExtensionB}

Reply:
Ok, yes i prefer the desirable option too. It is more flexible in case we need more attributes for the instantiation of an extension in the future.

Reply:
Unfortunately, I do not know how to write the XText grammar to do what we want. Maybe @dd4g12 can shed some light on this, but as far as I can see, it is not (yet) possible. So the options left for us are:
1. Keep the current EMF Containment and the current grammar. Due to the mandatory of `extensionID` attribute of `AbstractExtension`, we have to have the `contains` clause as follows
```
 contains
    a b : ExtensionA
 contains
    c d : ExtensionB
```
2. Change the EMF Containment to single `name` attribute and adapt the grammar accordingly. However, due the the `extensionID`, we will have something like this
```
 contains
    a : ExtensionA
 contains
    b : ExtensionA
 contains
    c: ExtensionB
 contains
    d: ExtensionB
```

Reply:
Note that extensionID is no longer mandatory (in the new prototype release of Event-B EMF)